File: frontend/src/api/rules.ts
import client from './client';
import type {
  RuleCatalogItem,
  RuleChangeLog,
  RuleDraft,
  RuleMergeResult,
  RuleRelation,
  RuleTemplate,
  RuleVersion,
  SimilarRuleCandidate,
  RuleStrictness,
  RuleSuggestion,
  RuleSuggestionReview,
} from '../types';

export async function getRuleCatalog(): Promise<RuleCatalogItem[]> {
  const res = await client.get<unknown[]>('/rules');
  return res.data.map(normalizeRuleCatalogItem);
}

export async function getRuleTemplates(): Promise<RuleTemplate[]> {
  const res = await client.get<unknown[]>('/rules/templates');
  return res.data.map(normalizeRuleTemplate);
}

export async function createRuleDraftFromTemplate(
  templateId: string,
  reason: string,
): Promise<RuleDraft> {
  const res = await client.post<unknown>(`/rules/templates/${templateId}/draft`, {
    reason,
  });
  return normalizeRuleDraft(res.data);
}

export async function updateRuleEnabled(
  ruleId: string,
  enabled: boolean,
  reason?: string,
): Promise<RuleCatalogItem> {
  const res = await client.put<unknown>(`/rules/${ruleId}/enabled`, {
    enabled,
    reason: reason || null,
  });
  return normalizeRuleCatalogItem(res.data);
}

export async function getRuleChangeLogs(limit = 100): Promise<RuleChangeLog[]> {
  const res = await client.get<unknown[]>('/rules/changes', {
    params: { limit },
  });
  return res.data.map(normalizeRuleChangeLog);
}

export async function getRuleRelations(ruleId: string): Promise<RuleRelation[]> {
  const res = await client.get<unknown[]>(`/rules/${ruleId}/relations`);
  return res.data.map(normalizeRuleRelation);
}

export async function getRuleVersions(ruleId: string): Promise<RuleVersion[]> {
  const res = await client.get<unknown[]>(`/rules/${ruleId}/versions`);
  return res.data.map(normalizeRuleVersion);
}

export async function updateCustomRule(
  ruleId: string,
  input: {
    name: string;
    rule_type: string;
    description: string;
    edit_reason: string;
  },
): Promise<RuleCatalogItem> {
  const res = await client.put<unknown>(`/rules/${ruleId}`, {
    name: input.name,
    rule_type: input.rule_type,
    description: input.description,
    edit_reason: input.edit_reason,
  });
  return normalizeRuleCatalogItem(res.data);
}

export async function rollbackCustomRule(
  ruleId: string,
  versionNo: number,
  reason: string,
): Promise<RuleCatalogItem> {
  const res = await client.put<unknown>(`/rules/${ruleId}/rollback`, {
    version_no: versionNo,
    reason,
  });
  return normalizeRuleCatalogItem(res.data);
}

export async function mergeCustomRule(
  sourceRuleId: string,
  targetRuleId: string,
  reason?: string,
): Promise<RuleMergeResult> {
  const res = await client.put<unknown>(`/rules/${sourceRuleId}/merge`, {
    target_rule_id: targetRuleId,
    reason: reason || null,
  });
  const item = res.data as Record<string, unknown>;
  return {
    source_rule_id: String(item.source_rule_id || ''),
    target_rule_id: String(item.target_rule_id || ''),
    relation_type: String(item.relation_type || ''),
    reason: item.reason == null ? null : String(item.reason),
    created_at: String(item.created_at || ''),
  };
}

export async function getRuleSuggestions(limit = 100): Promise<RuleSuggestion[]> {
  const res = await client.get<unknown[]>('/rules/suggestions', {
    params: { limit },
  });
  return res.data.map(normalizeRuleSuggestion);
}

export async function reviewRuleSuggestion(
  suggestionId: string,
  reviewStatus: RuleSuggestionReview['review_status'],
  reviewReason?: string,
): Promise<RuleSuggestionReview> {
  const res = await client.put<unknown>(`/rules/suggestions/${suggestionId}/review`, {
    review_status: reviewStatus,
    review_reason: reviewReason || null,
  });
  const item = res.data as Record<string, unknown>;
  return {
    suggestion_id: String(item.suggestion_id || ''),
    review_status:
      item.review_status === 'accepted' || item.review_status === 'rejected'
        ? item.review_status
        : 'pending',
    review_reason: item.review_reason == null ? null : String(item.review_reason),
  };
}

export async function getRuleDrafts(limit = 100): Promise<RuleDraft[]> {
  const res = await client.get<unknown[]>('/rules/drafts', {
    params: { limit },
  });
  return res.data.map(normalizeRuleDraft);
}

export async function reviewRuleDraft(
  draftId: number,
  draftStatus: 'published' | 'rejected' | 'pending_review',
  reviewReason?: string,
  options?: {
    similarRuleIds?: string[];
    differenceReason?: string;
  },
): Promise<RuleDraft> {
  const res = await client.put<unknown>(`/rules/drafts/${draftId}/review`, {
    draft_status: draftStatus,
    review_reason: reviewReason || null,
    similar_rule_ids: options?.similarRuleIds || null,
    difference_reason: options?.differenceReason || null,
  });
  return normalizeRuleDraft(res.data);
}

export async function updateRuleDraft(
  draftId: number,
  input: {
    name: string;
    rule_type: string;
    draft_content: string;
    edit_reason?: string;
  },
): Promise<RuleDraft> {
  const res = await client.put<unknown>(`/rules/drafts/${draftId}`, {
    name: input.name,
    rule_type: input.rule_type,
    draft_content: input.draft_content,
    edit_reason: input.edit_reason || null,
  });
  return normalizeRuleDraft(res.data);
}

export async function getSimilarRulesForDraft(
  draftId: number,
  limit = 10,
): Promise<SimilarRuleCandidate[]> {
  const res = await client.get<unknown[]>(`/rules/drafts/${draftId}/similar`, {
    params: { limit },
  });
  return res.data.map(normalizeSimilarRuleCandidate);
}

export async function reuseExistingRuleForDraft(
  draftId: number,
  targetRuleId: string,
  reason?: string,
): Promise<{
  draft_id: number;
  source_rule_id: string;
  target_rule_id: string;
  relation_type: string;
  reason: string | null;
}> {
  const res = await client.put<unknown>(`/rules/drafts/${draftId}/reuse`, {
    target_rule_id: targetRuleId,
    reason: reason || null,
  });
  const item = res.data as Record<string, unknown>;
  return {
    draft_id: Number(item.draft_id),
    source_rule_id: String(item.source_rule_id || ''),
    target_rule_id: String(item.target_rule_id || ''),
    relation_type: String(item.relation_type || ''),
    reason: item.reason == null ? null : String(item.reason),
  };
}

function normalizeRuleCatalogItem(raw: unknown): RuleCatalogItem {
  const item = raw as Record<string, unknown>;
  const strictness = item.strictness;

  return {
    id: String(item.id || ''),
    name: String(item.name || ''),
    domain: String(item.domain || ''),
    rule_type: String(item.rule_type || ''),
    strictness:
      strictness === 'balanced' || strictness === 'loose'
        ? (strictness as RuleStrictness)
        : 'strict',
    enabled: Boolean(item.enabled),
    keywords: Array.isArray(item.keywords)
      ? item.keywords.map((keyword) => String(keyword))
      : [],
    description: String(item.description || ''),
    source: String(item.source || ''),
    action: String(item.action || ''),
  };
}

function normalizeRuleTemplate(raw: unknown): RuleTemplate {
  const item = raw as Record<string, unknown>;
  const riskLevel = item.risk_level;
  return {
    id: String(item.id || ''),
    name: String(item.name || ''),
    category: String(item.category || ''),
    rule_type: String(item.rule_type || ''),
    risk_level:
      riskLevel === 'high' || riskLevel === 'medium' || riskLevel === 'low'
        ? riskLevel
        : String(riskLevel || ''),
    applicable_scene: String(item.applicable_scene || ''),
    evidence_requirements: Array.isArray(item.evidence_requirements)
      ? item.evidence_requirements.map((evidence) => String(evidence))
      : [],
    positive_examples: Array.isArray(item.positive_examples)
      ? item.positive_examples.map((example) => String(example))
      : [],
    negative_examples: Array.isArray(item.negative_examples)
      ? item.negative_examples.map((example) => String(example))
      : [],
    review_notes: String(item.review_notes || ''),
  };
}

function normalizeRuleChangeLog(raw: unknown): RuleChangeLog {
  const item = raw as Record<string, unknown>;
  return {
    id: Number(item.id),
    rule_id: String(item.rule_id || ''),
    previous_enabled: Boolean(item.previous_enabled),
    new_enabled: Boolean(item.new_enabled),
    reason: item.reason == null ? null : String(item.reason),
    created_at: String(item.created_at || ''),
  };
}

function normalizeRuleRelation(raw: unknown): RuleRelation {
  const item = raw as Record<string, unknown>;
  const direction = String(item.direction || '');
  return {
    id: Number(item.id),
    source_rule_id: String(item.source_rule_id || ''),
    target_rule_id: String(item.target_rule_id || ''),
    relation_type: String(item.relation_type || ''),
    reason: item.reason == null ? null : String(item.reason),
    created_at: String(item.created_at || ''),
    direction: direction === 'outgoing' || direction === 'incoming' ? direction : direction,
    related_rule_id: String(item.related_rule_id || ''),
    related_rule_name: String(item.related_rule_name || ''),
    related_rule_source: String(item.related_rule_source || ''),
    related_rule_status: String(item.related_rule_status || ''),
  };
}

function normalizeRuleVersion(raw: unknown): RuleVersion {
  const item = raw as Record<string, unknown>;
  return {
    id: Number(item.id),
    rule_id: String(item.rule_id || ''),
    version_no: Number(item.version_no || 0),
    name: String(item.name || ''),
    rule_type: String(item.rule_type || ''),
    description: String(item.description || ''),
    edit_reason: item.edit_reason == null ? null : String(item.edit_reason),
    created_at: String(item.created_at || ''),
  };
}

function normalizeRuleSuggestion(raw: unknown): RuleSuggestion {
  const item = raw as Record<string, unknown>;
  return {
    id: String(item.id || ''),
    source: String(item.source || ''),
    source_id: Number(item.source_id),
    suggestion_type: String(item.suggestion_type || ''),
    title: String(item.title || ''),
    reason: String(item.reason || ''),
    confidence: Number(item.confidence || 0),
    quality_status: String(item.quality_status || ''),
    quality_notes: item.quality_notes == null ? null : String(item.quality_notes),
    evidence_gaps: Array.isArray(item.evidence_gaps)
      ? item.evidence_gaps.map((gap) => String(gap))
      : [],
    requirement_id: Number(item.requirement_id),
    requirement_title: String(item.requirement_title || ''),
    requirement_content:
      item.requirement_content == null ? null : String(item.requirement_content),
    requirement_category:
      item.requirement_category == null ? null : String(item.requirement_category),
    tender_id: Number(item.tender_id),
    tender_title: item.tender_title == null ? null : String(item.tender_title),
    tender_filename:
      item.tender_filename == null ? null : String(item.tender_filename),
    evidence_snapshot:
      item.evidence_snapshot == null ? null : String(item.evidence_snapshot),
    review_status:
      item.review_status === 'accepted' || item.review_status === 'rejected'
        ? item.review_status
        : 'pending',
    review_reason: item.review_reason == null ? null : String(item.review_reason),
    created_at: String(item.created_at || ''),
  };
}

function normalizeRuleDraft(raw: unknown): RuleDraft {
  const item = raw as Record<string, unknown>;
  return {
    id: Number(item.id),
    source_suggestion_id: String(item.source_suggestion_id || ''),
    name: String(item.name || ''),
    rule_type: String(item.rule_type || ''),
    draft_content: String(item.draft_content || ''),
    draft_status: String(item.draft_status || ''),
    review_reason: item.review_reason == null ? null : String(item.review_reason),
    created_at: String(item.created_at || ''),
    updated_at: String(item.updated_at || ''),
  };
}

function normalizeSimilarRuleCandidate(raw: unknown): SimilarRuleCandidate {
  const item = raw as Record<string, unknown>;
  return {
    rule_id: String(item.rule_id || ''),
    name: String(item.name || ''),
    rule_type: String(item.rule_type || ''),
    source: String(item.source || ''),
    status: String(item.status || ''),
    similarity: Number(item.similarity || 0),
    reasons: Array.isArray(item.reasons)
      ? item.reasons.map((reason) => String(reason))
      : [],
    description: String(item.description || ''),
  };
}

File: backend/app/api/rules.py
"""规则库路由。"""

import logging

from fastapi import APIRouter
from pydantic import BaseModel

from app.services import rule_library_service
from app.utils.api_response import ErrorCode, error_response, success_response

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/rules", tags=["规则库"])


class RuleEnabledUpdate(BaseModel):
    enabled: bool
    reason: str | None = None


class RuleSuggestionReviewUpdate(BaseModel):
    review_status: str
    review_reason: str | None = None


class RuleDraftReviewUpdate(BaseModel):
    draft_status: str
    review_reason: str | None = None
    similar_rule_ids: list[str] | None = None
    difference_reason: str | None = None


class RuleDraftUpdate(BaseModel):
    name: str
    rule_type: str
    draft_content: str
    edit_reason: str | None = None


class RuleDraftReuseUpdate(BaseModel):
    target_rule_id: str
    reason: str | None = None


class RuleTemplateDraftCreate(BaseModel):
    reason: str | None = None


class RuleMergeUpdate(BaseModel):
    target_rule_id: str
    reason: str | None = None


class CustomRuleUpdate(BaseModel):
    name: str
    rule_type: str
    description: str
    edit_reason: str | None = None


class RuleRollbackUpdate(BaseModel):
    version_no: int
    reason: str | None = None


@router.get("", summary="获取规则库目录")
async def get_rule_catalog() -> dict:
    """返回当前系统内置规则目录。"""
    try:
        return success_response(data=await rule_library_service.list_rule_catalog_with_overrides())
    except Exception as e:
        logger.error("获取规则库失败: %s", str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"获取规则库失败: {str(e)}")


@router.get("/templates", summary="获取规则参考模板")
async def get_rule_templates() -> dict:
    """返回成熟标书分析场景中的参考模板，不直接作为生效规则。"""
    try:
        return success_response(data=rule_library_service.list_rule_templates())
    except Exception as e:
        logger.error("获取规则参考模板失败: %s", str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"获取规则参考模板失败: {str(e)}")


@router.post("/templates/{template_id}/draft", summary="从规则模板生成草案")
async def create_rule_draft_from_template(template_id: str, req: RuleTemplateDraftCreate) -> dict:
    """从参考模板生成待审核规则草案，不直接发布为生效规则。"""
    try:
        result = await rule_library_service.create_rule_draft_from_template(
            template_id,
            req.reason,
        )
        if result is None:
            return error_response(ErrorCode.PARAM_VALIDATION_FAILED, "模板不存在或使用原因缺失")
        return success_response(data=result, message="规则模板已生成待审核草案")
    except Exception as e:
        logger.error("从规则模板生成草案失败: template_id=%s, error=%s", template_id, str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"从规则模板生成草案失败: {str(e)}")


@router.get("/{rule_id}/relations", summary="获取规则关系")
async def get_rule_relations(rule_id: str) -> dict:
    """返回某条规则的复用、相似等追溯关系。"""
    try:
        return success_response(data=await rule_library_service.list_rule_relations(rule_id))
    except Exception as e:
        logger.error("获取规则关系失败: rule_id=%s, error=%s", rule_id, str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"获取规则关系失败: {str(e)}")


@router.get("/{rule_id}/versions", summary="获取规则版本历史")
async def get_rule_versions(rule_id: str) -> dict:
    """返回自定义规则的历史版本。"""
    try:
        return success_response(data=await rule_library_service.list_rule_versions(rule_id))
    except Exception as e:
        logger.error("获取规则版本历史失败: rule_id=%s, error=%s", rule_id, str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"获取规则版本历史失败: {str(e)}")


@router.put("/{rule_id}/merge", summary="归并自定义规则")
async def merge_custom_rule(rule_id: str, req: RuleMergeUpdate) -> dict:
    """将一条重复自定义规则归并到另一条已存在规则，并停用源规则。"""
    try:
        result = await rule_library_service.merge_custom_rule(
            rule_id,
            req.target_rule_id,
            req.reason,
        )
        if result is None:
            return error_response(ErrorCode.PARAM_VALIDATION_FAILED, "源规则或目标规则不存在，或源规则不可归并")
        return success_response(data=result, message="规则已归并")
    except Exception as e:
        logger.error("归并规则失败: rule_id=%s, error=%s", rule_id, str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"归并规则失败: {str(e)}")


@router.put("/{rule_id}/rollback", summary="回滚自定义规则")
async def rollback_custom_rule(rule_id: str, req: RuleRollbackUpdate) -> dict:
    """将自定义规则恢复到某个历史版本，并保存回滚前内容。"""
    try:
        result = await rule_library_service.rollback_custom_rule(
            rule_id,
            req.version_no,
            req.reason,
        )
        if result is None:
            return error_response(ErrorCode.PARAM_VALIDATION_FAILED, "规则不存在、版本不存在或回滚原因缺失")
        return success_response(data=result, message="规则已回滚")
    except Exception as e:
        logger.error("回滚自定义规则失败: rule_id=%s, error=%s", rule_id, str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"回滚自定义规则失败: {str(e)}")


@router.put("/{rule_id}", summary="编辑自定义规则")
async def update_custom_rule(rule_id: str, req: CustomRuleUpdate) -> dict:
    """编辑已发布自定义规则，并保存修改前版本。"""
    try:
        result = await rule_library_service.update_custom_rule(
            rule_id,
            req.name,
            req.rule_type,
            req.description,
            req.edit_reason,
        )
        if result is None:
            return error_response(ErrorCode.PARAM_VALIDATION_FAILED, "规则不存在、不是自定义规则或修改原因缺失")
        return success_response(data=result, message="规则已更新")
    except Exception as e:
        logger.error("编辑自定义规则失败: rule_id=%s, error=%s", rule_id, str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"编辑自定义规则失败: {str(e)}")


@router.put("/{rule_id}/enabled", summary="更新规则启停状态")
async def update_rule_enabled(rule_id: str, req: RuleEnabledUpdate) -> dict:
    """更新单条内置规则启停状态。"""
    try:
        result = await rule_library_service.update_rule_enabled(
            rule_id,
            req.enabled,
            req.reason,
        )
        if result is None:
            return error_response(ErrorCode.PARAM_VALIDATION_FAILED, "规则不存在")
        return success_response(data=result, message="规则状态已更新")
    except Exception as e:
        logger.error("更新规则启停失败: rule_id=%s, error=%s", rule_id, str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"更新规则失败: {str(e)}")


@router.get("/changes", summary="获取规则变更记录")
async def get_rule_change_logs(limit: int = 100) -> dict:
    """返回规则启停变更记录。"""
    try:
        return success_response(data=await rule_library_service.list_rule_change_logs(limit))
    except Exception as e:
        logger.error("获取规则变更记录失败: %s", str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"获取规则变更记录失败: {str(e)}")


@router.get("/suggestions", summary="获取规则候选建议")
async def get_rule_suggestions(limit: int = 100) -> dict:
    """从人工错例中生成规则候选建议。"""
    try:
        return success_response(data=await rule_library_service.list_rule_suggestions(limit))
    except Exception as e:
        logger.error("获取规则候选建议失败: %s", str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"获取规则候选建议失败: {str(e)}")


@router.put("/suggestions/{suggestion_id}/review", summary="处理规则候选建议")
async def review_rule_suggestion(
    suggestion_id: str,
    req: RuleSuggestionReviewUpdate,
) -> dict:
    """记录规则建议的采纳/忽略状态。"""
    try:
        result = await rule_library_service.review_rule_suggestion(
            suggestion_id,
            req.review_status,
            req.review_reason,
        )
        if result is None:
            return error_response(ErrorCode.PARAM_VALIDATION_FAILED, "建议处理状态无效")
        return success_response(data=result, message="规则建议处理状态已更新")
    except Exception as e:
        logger.error("处理规则候选建议失败: suggestion_id=%s, error=%s", suggestion_id, str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"处理规则建议失败: {str(e)}")


@router.get("/drafts", summary="获取规则草案")
async def get_rule_drafts(limit: int = 100) -> dict:
    """返回由已采纳建议生成的规则草案。"""
    try:
        return success_response(data=await rule_library_service.list_rule_drafts(limit))
    except Exception as e:
        logger.error("获取规则草案失败: %s", str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"获取规则草案失败: {str(e)}")


@router.put("/drafts/{draft_id}", summary="编辑规则草案")
async def update_rule_draft(draft_id: int, req: RuleDraftUpdate) -> dict:
    """编辑待审核规则草案，已发布/已驳回草案不可直接修改。"""
    try:
        result = await rule_library_service.update_rule_draft(
            draft_id,
            req.name,
            req.rule_type,
            req.draft_content,
            req.edit_reason,
        )
        if result is None:
            return error_response(ErrorCode.PARAM_VALIDATION_FAILED, "规则草案不存在、已发布或内容无效")
        return success_response(data=result, message="规则草案已更新")
    except Exception as e:
        logger.error("编辑规则草案失败: draft_id=%s, error=%s", draft_id, str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"编辑规则草案失败: {str(e)}")


@router.get("/drafts/{draft_id}/similar", summary="检测规则草案相似规则")
async def get_similar_rules_for_draft(draft_id: int, limit: int = 10) -> dict:
    """发布草案前检测相似内置规则、自定义规则和其他草案。"""
    try:
        result = await rule_library_service.list_similar_rules_for_draft(draft_id, limit)
        if result is None:
            return error_response(ErrorCode.PARAM_VALIDATION_FAILED, "规则草案不存在")
        return success_response(data=result)
    except Exception as e:
        logger.error("检测相似规则失败: draft_id=%s, error=%s", draft_id, str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"检测相似规则失败: {str(e)}")


@router.put("/drafts/{draft_id}/reuse", summary="复用已有规则")
async def reuse_existing_rule_for_draft(draft_id: int, req: RuleDraftReuseUpdate) -> dict:
    """将待审核草案关联到已有规则，避免重复发布。"""
    try:
        result = await rule_library_service.reuse_existing_rule_for_draft(
            draft_id,
            req.target_rule_id,
            req.reason,
        )
        if result is None:
            return error_response(ErrorCode.PARAM_VALIDATION_FAILED, "规则草案或目标规则不存在，或草案状态不可复用")
        return success_response(data=result, message="已复用已有规则")
    except Exception as e:
        logger.error("复用已有规则失败: draft_id=%s, error=%s", draft_id, str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"复用已有规则失败: {str(e)}")


@router.put("/drafts/{draft_id}/review", summary="审核规则草案")
async def review_rule_draft(draft_id: int, req: RuleDraftReviewUpdate) -> dict:
    """发布或驳回规则草案。"""
    try:
        result = await rule_library_service.review_rule_draft(
            draft_id,
            req.draft_status,
            req.review_reason,
            req.similar_rule_ids,
            req.difference_reason,
        )
        if result is None:
            return error_response(ErrorCode.PARAM_VALIDATION_FAILED, "规则草案不存在或状态无效")
        return success_response(data=result, message="规则草案审核状态已更新")
    except Exception as e:
        logger.error("审核规则草案失败: draft_id=%s, error=%s", draft_id, str(e))
        return error_response(ErrorCode.INTERNAL_SERVER_ERROR, f"审核规则草案失败: {str(e)}")
